### PR TITLE
High-level motivation for operator-based reconciliation

### PR DESCRIPTION
**Pull request by jakobmoellersap**
**Description**

Changes proposed in this pull request:

- Creation of Operator-based Reconciliation and Replacement of the current Mothership Reconciler
- Multi-Path Implementation of Kyma Deployments from Small to Enterprise Scale with different Configurations

**Related issue(s)**

- https://github.com/kyma-project/kyma/issues/13759
- https://github.com/kyma-project/community/tree/main/concepts/infrastructure-reconciliation

### CODE REVIEW

**Review by tobiscr — APPROVED** [review]

File: concepts/operator-reconciliation/README.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+kubectl or API Interaction as easy as possible. The easiest way for us to allow this is by leveraging the kubernetes 
+built-in `status` field besides the spec to also track reconciliations. This allows us to have eventually consistent 
+clusters that do not rely on any worker queue to get processed but instead can be processed on demand by any free 
+operator taking care of the Cluster reconciliation.
+
+#### How do we protect the changes in the cluster from a user?
+
+While there are generally multiple approaches for this in a production cluster (e.g. limiting dangerous access to 
+`kyma-system` to SREs or running a dedicated control plane), we also will make use of admission hooks, validation 
+hooks as well as finalizers to enable a resilient interaction with the cluster CRs. We will avoid and stop any
```
Comment (adityabhatia) [inline concepts/operator-reconciliation/README.md:157]:
```suggestion
`kyma-system` to SREs or running a dedicated control plane), we also will make use of admission controllers (mutations / validation webhooks) as well as finalizers to enable a resilient interaction with the cluster CRs. We will avoid and stop any 
```

File: concepts/operator-reconciliation/README.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+## Motivation
+
+Reconciler is a framework to install, update, and repair Kyma components in managed Kyma Runtimes. The reconciliation runs in the loop to ensure that all components are up and running and properly configured.
+While we currently have a custom reconciliation logic in place, we could make use of the Kubernetes Control Loop and reconciliation and an Operator-based approach to maintain our reconciliation through Kubernetes natively.
+
+We are currently having these issues in our current reconciliation:
+
+- Scheduling / Job-based
+  - We have inbuilt limits on how fast a reconciliation can work by having bad worst case scheduling scenarios (e.g. reconciliation updates only get picked up every 45 seconds through our bookkeeper)
+  - Reconciliation must be externally managed from outside the Cluster as there is no internal state management inside a shoot, this requires either the CLI or a control plane cluster with a Mothership
```
Comment (adityabhatia) [inline concepts/operator-reconciliation/README.md:12]: Using the Kyma CLI you can also use it on your own cluster as well, also private gardener shoots?

File: concepts/operator-reconciliation/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+From the [Kubernetes Documentation](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/):
+
+> Kubernetes is designed for automation. Out of the box, you get lots of built-in automation from the core of Kubernetes. You can use Kubernetes to automate deploying and running workloads, and you can automate how Kubernetes does that.
+> Kubernetes' operator pattern concept lets you extend the cluster's behaviour without modifying the code of Kubernetes itself by linking controllers to one or more custom resources. Operators are clients of the Kubernetes API that act as controllers for a Custom Resource.
+
+We want to make use of this design pattern to optimize and enhance our current Reconciliation control loop. The core characterics are:
+
+- The new architecture should encompass one or many Operator(s) which can reconcile a Gardener Shoot Cluster from empty to working Kyma on their own without external influence
+- The architecture uses CRDs to manage Kyma Components and their necessary Reconciliation
```
Comment (adityabhatia) [inline concepts/operator-reconciliation/README.md:37]:
```suggestion
- The architecture uses versioned configurations (ConfigMaps) to manage component custom resources, which should trigger component reconciliations on the cluster
```

File: concepts/operator-reconciliation/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+- The architecture uses CRs to report State of the Reconciliation of a Cluster that can be externally viewed via Kubernetes API (e.g. through kubectl or a central provisioning component in the control plane)
+- The reconciliation infrastructure can be simply installed through provisioning of CRDs and/or Operators, maybe through its own Helm Chart
```
Comment (adityabhatia) [inline concepts/operator-reconciliation/README.md:39]:
```suggestion
- The reconciliation infrastructure can be simply installed through provisioning of CRDs, component configurations and operators, maybe through its own Helm Chart
```

File: concepts/operator-reconciliation/README.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+- The reconciliation does not need explicit State Management outside of CRs managed by the Operator like we have currently with the Mothership
+- The components for reconciliation can be deployed inside the Cluster to reconcile and manage itself
+- The reconciliation components can be integrated through a reported Cluster State into an external Provisioning / Reconciliation for consolidation in a Control Plane Cluster (Kubeception)
+- The reconciliation uses one single source of truth for configuring which Kyma Components are enabled and what Settings are used in the Cluster
+- The reconciliation only consumes measurable load when actively reconciling by smartly managing deployments and instances used for reconciliation, scaling down and passively watching the compliant cluster
+- The reconciliation is able to be debugged and developed by multiple teams for multiple components without strong coupling effects
```
Comment (adityabhatia) [inline concepts/operator-reconciliation/README.md:45]:
```suggestion
- The reconciliations of registered components are de-coupled, thereby providing a better root cause analysis
```

File: concepts/operator-reconciliation/README.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+status, which also means that resources should never be exposed to an untrusted source in order to ensure safe 
+reconciliation.
+
+#### One CRD vs Multiple CRDs for maintaining Reconciliation Configuration?
+
+While we will maintain One CRD for the Kyma Installation managed by the `kyma-operator`, we will also make use of a 
+generic Approach to integrate components dynamically based on a template. This template is a base for any component 
+that needs to be reconciled by our control loop. We will explicitly NOT have hard-coded CRDs for Components and want 
+to dynamically enhance a Kyma Control Plane with Components without having to deliver a new Operator Binary to the 
+Control Plane.
```
[the remark this answers is not in the document]
Comment (adityabhatia) [reply concepts/operator-reconciliation/README.md:168]: there's some formatting issue as well here

File: concepts/operator-reconciliation/README.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
+# Operator-based Reconciler
+
+## Motivation
+
+Reconciler is a framework to install, update, and repair Kyma components in managed Kyma Runtimes. The reconciliation runs in the loop to ensure that all components are up and running and properly configured.
+While we currently have a custom reconciliation logic in place, we could make use of the Kubernetes Control Loop and reconciliation and an Operator-based approach to maintain our reconciliation through Kubernetes natively.
+
```
[the remark this answers is not in the document]
Comment (NHingerl) [reply concepts/operator-reconciliation/README.md:7]: The "(s)" plural is not recommended. I suggest "...make use of one or more Kubernetes Operators..." instead.

File: concepts/operator-reconciliation/README.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+We are currently having these issues in our current reconciliation:
+
+- Scheduling / Job-based
+  - We have inbuilt limits on how fast a reconciliation can work by having bad worst case scheduling scenarios (e.g. reconciliation updates only get picked up every 45 seconds through our bookkeeper)
```
[the remark this answers is not in the document]
Comment (NHingerl) [reply concepts/operator-reconciliation/README.md:11]:
```suggestion
  -   - Static limits on how fast a reconciliation can achieve its desired state, by having worst-case scheduling scenarios (for example, reconciliation updates are picked up only every 45 seconds through our bookkeeper).
```

File: concepts/operator-reconciliation/README.md (hunk 1 above)
Comment (NHingerl) [reply concepts/operator-reconciliation/README.md:157]: The last sentence seems incomplete.

**Review by tobiscr — APPROVED** [review]

File: concepts/operator-reconciliation/README.md (hunk 2 above)
Comment (jakobmoellersap) [reply concepts/operator-reconciliation/README.md:12]: That would have been my expectation, I still think you will have a provision and deploy command. So if you already have your private gardener shoot, you only need to deploy with the kubeconfig

File: concepts/operator-reconciliation/README.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+For Operator-based reconciliations, we want to enable maximum flexibility of the architecture depending on the needs of the Kyma user. This means that we want to allow
+
+1. Fully Independent Small-Scale Clusters able to reconcile themselves, relying on external node setup (e.g. running for a local evaluation in k3d or for smaller scale deployments)
+2. Fully centralized Control Plane Configurations, in which the Control Plane is not only provisioning the cluster, but also hosting the reconciliation, ensuring minimal overhead in the workload cluster
+3. Lightweight Control Plane Configurations (Hybrid Mode), in which the Control Plane is able to provision the cluster and create/update an inventory of all managed cluster, but not taking care of reconciliation
+
+The main focus of the POC will be to prove the functionality of the Reconciliation Operator to work with all 3 modes independant of the final implementation. The POC needs to prove that we are able to reconcile a cluster remotely or in-cluster without a significant difference in reconciliation logic.
+
+![high_level_view_operator-reconciliation](assets/high_level_view_operator-reconciliation.png)
+
```
Comment (tobiscr) [inline concepts/operator-reconciliation/README.md:79]: Like the diagram .. but would be good to visualise also the Kyma CR (KEB is not talking to kyma-operator directly)

Would be great to add also a short table below the diagram describing each visualised component with 1-2 sentences what it is/what its purpose is

File: concepts/operator-reconciliation/README.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+### Single-Cluster (Small-Scale)
+
+![single_cluster_configuration](assets/single_cluster_configuration.png)
```
Comment (tobiscr) [inline concepts/operator-reconciliation/README.md:82]: Good to point out these two different modes: small-scale and centralised. But I could imaging that we should put both in one diagram to make also obvious, that the differences are not that huge:

* The main difference is, that in the centalised deployment will be many Kyma-CRs, Cluster-CRs etc. But in the small-scale mode will exist only one of them pointing to the local cluster.

Comment (jakobmoellersap) [reply concepts/operator-reconciliation/README.md:84]: Big bad diagram will contain only one visualization with a disclaimer for small-scale

File: concepts/operator-reconciliation/README.md (hunk 9 above)
Comment (jakobmoellersap) [reply concepts/operator-reconciliation/README.md:79]: Will take a bit but will do!

Comment (NHingerl) [reply concepts/operator-reconciliation/README.md:79]: FYI, we have a guideline for diagrams: https://kyma-project.io/community/guidelines/content-guidelines/09-diagrams/#examples

Comment (jakobmoellersap) [reply concepts/operator-reconciliation/README.md:79]: @NHingerl  While I saw the guideline, I was using the Reference Diagramming language of SAP https://help.sap.com/docs/SAP_POWERDESIGNER/1cc460ad80f446e6a9d19303919ee269/c818cfa96e1b1014abb5d137d4620b1e.html?version=16.6.4&locale=en-US
http://www.fmc-modeling.org/download/fmc-and-tam/SAP-TAM_Standard.pdf

This is afaik the only widely accepted architecture standard @ SAP. If this is too far off for modelling kyma diagrams I think we will have to either rework these guidelines or add to them a mapping from TAM.

Comment (NHingerl) [reply concepts/operator-reconciliation/README.md:79]: As discussed, I think the SAP style guide you used is better suited to the needs of your document. 👍

**Comment by NHingerl** [comment]
Decision: 
Separate docs for motivation (this one) and for explaining the status quo (new doc).

**Review by tobiscr — APPROVED** [review]

**Comment by jakobmoellersap** [comment]
@NHingerl could we merge this and then iterate on it? That way its easier to change our high level motivation after we updated the rest of the docs

**Pull request merged**